rust/src/utils/tfrecord.rs

```diff
@@ -0,0 +1,777 @@
+                            )
+                        } else {
+                            values.append_slice(tensor.float_val.as_slice());
+                        }
+                    }
+                } else {
+                    values.append_nulls(list_size);
+                }
+            }
+            Arc::new(values.finish())
+        }
+        DataType::Float64 => {
+            let mut values = Float64Builder::with_capacity(features.len());
+            for tensors in tensor_iter {
+                if let Some(tensors) = tensors? {
+                    for tensor in tensors {
+                        validate_tensor(&tensor, type_info)?;
+                        if tensor.float_val.is_empty() {
+                            append_primitive_from_slice(
+                                &mut values,
+                                tensor.tensor_content.as_slice(),
+                                |bytes| f64::from_le_bytes(bytes.try_into().unwrap()),
+                            )
+                        } else {
+                            values.append_slice(tensor.double_val.as_slice())
+                        };
+                    }
+                } else {
+                    values.append_nulls(list_size);
+                }
+            }
+            Arc::new(values.finish())
+        }
+        _ => Err(Error::IO {
+            message: format!("unsupported type {:?}", type_info.leaf_type),
+        })?,
+    };
+
+    Ok((values, offsets))
+}
+
+fn validate_tensor(tensor: &TensorProto, type_info: &TypeInfo) -> Result<()> {
+    let tensor_shape = tensor.tensor_shape.as_ref().unwrap();
+    let length = tensor_shape.dim.iter().map(|d| d.size as i32).product();
+    if type_info.fsl_size != Some(length) {
+        return Err(Error::IO {
+            message: format!(
+                "tensor length mismatch: expected {}, got {}",
+                type_info.fsl_size.unwrap(),
+                length
+            ),
+        });
+    }
+
+    let data_type = tensor_dtype_to_arrow(&tensor.dtype())?;
+    if data_type != type_info.leaf_type {
+        return Err(Error::IO {
+            message: format!(
+                "tensor type mismatch: expected {:?}, got {:?}",
+                type_info.leaf_type,
+                tensor.dtype()
+            ),
+        });
+    }
+
+    Ok(())
+}
+
+/// Given a potentially unaligned slice, append the slice to the builder.
+fn append_primitive_from_slice<T>(
+    builder: &mut PrimitiveBuilder<T>,
+    slice: &[u8],
+    parse_val: impl Fn(&[u8]) -> T::Native,
+) where
+    T: arrow::datatypes::ArrowPrimitiveType,
+{
+    let (prefix, middle, suffix) = unsafe { slice.align_to::<T::Native>() };
+    for val in prefix.chunks_exact(T::get_byte_width()) {
+        builder.append_value(parse_val(val));
+    }
+
+    builder.append_slice(middle);
+
+    for val in suffix.chunks_exact(T::get_byte_width()) {
+        builder.append_value(parse_val(val));
+    }
+}
```

Comment:
Just out of curiosity, what's going on here?  What's the general policy for unsafe blocks in lance?

Reply:
Oh this is to convert from a `&[u8]` (slice of bytes) to a `&[T::Native]` (`f32` for a Float32Array, `i64` for a Int64Array, etc.). The slice of bytes isn't guaranteed to be aligned, so this breaks it up into three pieces: the unaligned prefix, an aligned middle part, and the unaligned suffix.

> What's the general policy for unsafe blocks in lance?
We haven't established a policy, but one we could adopt is: when ever you have an unsafe block, add a comment above explaining why the next line is sound.

Here, the soundness of `align_to` depends on whether the bytes that form the middle piece are actually valid values for the `T::Native` type. I'm not 100% certain, but I don't think this applies to f32 or i64. Are there any 32-bit values that aren't a valid f32? Or 64-bit values that aren't a valid i64? I could see this being an issue for boolean values represented with a byte; there's only two valid values but 256 possible values, so 244 possible invalid values.

Reply:
I see now, I just read up on align_to.  It's a nice convenience :).  The only other thing I could think of would be endianness.  For the `half_val` stuff we are fortunate that protobuf will handle endianess for us.  However, for the `tensor_content` option it looks like tensorflow is just reinterpret casting the data bytes.  So if the tfrecord file is created on a machine with one kind of endianess and then read on a machine with a different kind of endianess you will get back garbage.

However, maybe just a comment / to-do PR and we can worry about it later when we have users / use cases for big endian machines.